query.go

```diff
@@ -187,7 +187,7 @@ func (f *field) tokenize(w io.Writer, fields Fields) (bool, error) { //nolint:go
 
 	if f.Keep || fields == nil {
 		write = true
-	} else if desired, exists := fields[f.Decl.Name]; exists {
+	} else if desired, exists := fields[f.Decl.Name]; exists { // this is probably the offending line
```

Comment:
@grevych admittedly, i didn't spend too much time digging into this, but i do have a solid hunch that this line is the reason this is happening.

it's looking for `f.Decl.Name` in the fields map, which is defined on line 282. its just name of the field with no notion of whether or not it is a parent/child etc.

so when a parent and a child share the same field name, and one of them requests it (there is a chance this only happens in one direction, if the parent requests the field name a child has, but not vice-versa), both of them get the field because of this "simple" map lookup.

i hope this is making a little bit of sense, and I'm happy to help a little more if need-be.